File: app/src/main/java/io/legado/app/ui/book/source/manage/SourceSorter.kt
package io.legado.app.ui.book.source.manage

import io.legado.app.data.entities.BookSource
import io.legado.app.data.entities.BookSourcePart
import io.legado.app.utils.cnCompare

object SourceSorter {

    fun sort(
        data: List<BookSourcePart>,
        mode: BookSourceSort,
        ascending: Boolean
    ): List<BookSourcePart> {
        val result = when (mode) {
            BookSourceSort.Default -> data.sortedBy { it.customOrder }
            BookSourceSort.Name -> data.sortedWith { a, b ->
                a.bookSourceName.cnCompare(b.bookSourceName)
            }
            BookSourceSort.Url -> data.sortedBy { it.bookSourceUrl }
            BookSourceSort.Weight -> data.sortedBy { it.weight }
            BookSourceSort.Update -> data.sortedByDescending { it.lastUpdateTime }
            BookSourceSort.Respond -> data.sortedBy { it.respondTime }
            BookSourceSort.Enable -> data.sortedWith { a, b ->
                var r = -a.enabled.compareTo(b.enabled)
                if (r == 0) r = a.bookSourceName.cnCompare(b.bookSourceName)
                r
            }
        }
        return if (ascending) result else result.reversed()
    }

    @JvmName("sortSources")
    fun sort(
        data: List<BookSource>,
        mode: BookSourceSort,
        ascending: Boolean
    ): List<BookSource> {
        val result = when (mode) {
            BookSourceSort.Default -> data.sortedBy { it.customOrder }
            BookSourceSort.Name -> data.sortedWith { a, b ->
                a.bookSourceName.cnCompare(b.bookSourceName)
            }
            BookSourceSort.Url -> data.sortedBy { it.bookSourceUrl }
            BookSourceSort.Weight -> data.sortedBy { it.weight }
            BookSourceSort.Update -> data.sortedByDescending { it.lastUpdateTime }
            BookSourceSort.Respond -> data.sortedBy { it.respondTime }
            BookSourceSort.Enable -> data.sortedWith { a, b ->
                var r = -a.enabled.compareTo(b.enabled)
                if (r == 0) r = a.bookSourceName.cnCompare(b.bookSourceName)
                r
            }
        }
        return if (ascending) result else result.reversed()
    }
}

File: modules/web/src/utils/sourceSorter.ts
import type { Source } from '@/source'

export type SortMode = 'default' | 'name' | 'url' | 'weight' | 'update' | 'respond' | 'enable'

export function sortSources(sources: Source[], mode: SortMode, ascending: boolean): Source[] {
  const result = [...sources]
  switch (mode) {
    case 'default':
      result.sort((a, b) => a.customOrder - b.customOrder)
      break
    case 'name':
      result.sort((a, b) => a.bookSourceName.localeCompare(b.bookSourceName, 'zh-CN-u-co-pinyin'))
      break
    case 'url':
      result.sort((a, b) => (a.bookSourceUrl || '').localeCompare(b.bookSourceUrl || ''))
      break
    case 'weight':
      result.sort((a, b) => a.weight - b.weight)
      break
    case 'update':
      result.sort((a, b) => b.lastUpdateTime - a.lastUpdateTime)
      break
    case 'respond':
      result.sort((a, b) => a.respondTime - b.respondTime)
      break
    case 'enable':
      result.sort((a, b) => {
        if (a.enabled !== b.enabled) return Number(b.enabled) - Number(a.enabled)
        return a.bookSourceName.localeCompare(b.bookSourceName, 'zh-CN-u-co-pinyin')
      })
      break
  }
  return ascending ? result : result.reverse()
}
